### ISSUE TITLE
Decoder input

### ISSUE THREAD

**Issue by mun94** [jadore801120/attention-is-all-you-need-pytorch#8]
Hi, I am not sure if you are feeding the right input to the decoder.

(pg. 2) "Given z, the decoder then generates an output sequence (y<sub>1</sub>, ..., y<sub>m</sub>) of symbols one element at a time. At each step the model is auto-regressive, consuming the previously generated symbols as additional input when generating the next."

I believe your decoder input is a batch of target sequences.

**Comment by jadore801120** [comment]
Hi @mun94 ,

I think it is not clear that this paragraph is talking about training phase or inference phase.

Therefore, I follow the classic Recurrent Sequence to Sequence training style.
The method of feeding the target sequence into the decoder during training phase is called "teacher forcing", which can make the model converge faster.
Training with the last output from decoder may also works well, but may take longer time to train.
In the inference phase, I use the last output from the decoder as its next input.

However, I may be wrong.
If you found out that the authors do train without teacher forcing, please let me know.
Thank you!

Yu-Hsiang

**Comment by mun94** [comment]
Yeah it was not clear to be as well, but when I looked at their tensorflow implementation, I realized that their decoder works like word-based RNN. They feed back the output word token and append the output word token to the encoder input at each time with the shift.

For example, the input at the very first input looks like y<sub>t = 0</sub> = [[0, 0, 0, 0, 2], [0, 0, 0, 0, 2]] (batch=2, max_seqlen=5, '2'='&lt;s&gt;']. Then, once you get the output ('4'='hi') from the decoder, it gets fed back into the input sequence and shift by one, which results in y<sub>1</sub> = [[0, 0, 0, 2, 4], [0, 0, 0, 2, 4]].

**Comment by mun94** [comment]
And, wow, 'teacher forcing' seems powerful.

1) How do you deploy the model? In other words, during testing/deployment (not validation), what do you put in replace of target sequences?

2) Would you mind linking the paper that you are referenced?

Thank you!

**Comment by jadore801120** [comment]
Hi @mun94 ,

Would you mind to provide the link and the **exact** code part about the tensorflow implementation you mentioned?
I haven't have time to dig into the official tf code.

**Comment by mun94** [comment]
I believe this does it. Let me know if I am wrong :) [https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py#L227](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py#L227)

**Comment by jadore801120** [comment]
Hi @mun94 ,

The teacher forcing concept are first named in [A Learning Algorithm for Continually Running Fully
Recurrent Neural Networks ](http://www.mitpressjournals.org/doi/pdf/10.1162/neco.1989.1.2.270).
However, it has been well described in the introduction part in [Professor Forcing: A New Algorithm for Training Recurrent Networks](https://papers.nips.cc/paper/6099-professor-forcing-a-new-algorithm-for-training-recurrent-networks.pdf).

**Comment by jadore801120** [comment]
Hi @mun94 ,

I am sorry. I did not see where the decoder take the last output instead of ground truth in your link.
The part you post seems only padding the data instances to the same length for mini batch operation.

I am not an expert of TensorFlow, so please highlight **the exact code statements** for me instead of just the function name. Thank you.

**Comment by mun94** [comment]
I am not an expert either so I am not sure if my explanation will make sense for you lol.

But basically, in the function [below](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/t2t_model.py#L270)
`def infer_step(recent_output, _)`,
you it does extra padding and add the 'recent output' (line 273):
`padded = tf.pad(recent_output, [[0, 0], [0, 1], [0, 0], [0, 0]])`.
Then they slide the window of the decoder input (line 305-309):
`result = tf.foldl(  
        infer_step,  
        tf.range(decode_length),  
        initializer=initial_output,  
        back_prop=False,  
        parallel_iterations=1)`

where 'decode_length=50' (line 140) is the window size.

They do it token-wise. I guess you are giving the whole sequence (which you referred as the 'last output from the decoder') into the decoder, right?

**Comment by jadore801120** [comment]
Excuse me.

> I guess you are giving the whole sequence (which you referred as the 'last output from the decoder') into the decoder, right?

Yes, I did. 
And I also said that the teacher forcing (whole target seq into the decoder) is only used in **training** phase.

The code you paste is under the function called `infer` and it is for **inference** phase.
In the inference phase, I do the exact same thing as they did.

I think we are discussing about the teacher forcing method in the training phase.

**Comment by mun94** [comment]
Sorry, I just wasn't clear about the term inference phase.

**Comment by jadore801120** [comment]
The inference phase is simply the prediction phase (or test phase).
You could take a look on the following resources to get more background knowledge.

[Tensorflow tutorial](https://www.tensorflow.org/tutorials/seq2seq): step by step, with original paper.
[OpenNMT-py](https://github.com/OpenNMT/OpenNMT-py): very well-written, a little bit complex


Since the misunderstanding is cleared up, let me close this issue.

**Comment by mun94** [comment]
Can you please send me the link to your code where the decoder behaves in a regressive manner? THank you :)

**Comment by jadore801120** [comment]
The code is here: https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/master/transformer/Translator.py#L81-L102

This part also include the beam search part (not a greedy infer), so maybe hard to read.

Long story short. 
The decode output is stored in beam `b`.
And the last output from decoder are retrieved with the `get_current_state` and take as the next input.
https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/master/transformer/Translator.py#L81-L82 .

**Comment by ZedYeung** [comment]
Why not just set teacher forcing probability as spro did.

**Comment by ylmeng** [comment]
It might be good to stop using teacher forcing after certain epochs, if we can put such a hyperparameter in the scheduler.

**Comment by tonyhu20116543** [comment]
@jadore801120 
Hi guy, you are very nice to provide the pytorch code of transformer, great work!
But I have a question for you, do you implement the teacher-forcing strategy in your training process below?
[https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/master/transformer/Models.py#L204-L206](url)
It seems you did not use the teacher forcing in the training, right?
